diff --git a/linker/linker.cpp b/linker/linker.cpp
index 54867dc..24b5f58 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -63,6 +63,7 @@
  *
  * open issues / todo:
  *
+ * - are we doing everything we should for ARM_COPY relocations?
  * - cleaner error reporting
  * - after linking, set as much stuff as possible to READONLY
  *   and NOEXEC
@@ -1271,18 +1272,53 @@ int soinfo::Relocate(ElfW(Rela)* rela, unsigned count) {
         *reinterpret_cast<ElfW(Addr)*>(reloc) = call_ifunc_resolver(base + rela->r_addend);
         break;
 
-      case R_AARCH64_COPY:
-        /*
-         * ET_EXEC is not supported so this should not happen.
-         *
-         * http://infocenter.arm.com/help/topic/com.arm.doc.ihi0044d/IHI0044D_aaelf.pdf
-         *
-         * Section 4.7.1.10 "Dynamic relocations"
-         * R_AARCH64_COPY may only appear in executable objects where e_type is
-         * set to ET_EXEC.
-         */
-        DL_ERR("%s R_AARCH64_COPY relocations are not supported", name);
-        return -1;
+      case R_AARCH64_COPY: 
+     if ((flags & FLAG_EXE) == 0) {
+         /*
+           * http://infocenter.arm.com/help/topic/com.arm.doc.ihi0044d/IHI0044D_aaelf.pdf
+           *
+           * Section 4.7.1.10 "Dynamic relocations"
+           * R_AARCH64_COPY may only appear in executable objects where e_type is
+           * set to ET_EXEC.
+           *
+           * FLAG_EXE is set for both ET_DYN and ET_EXEC executables.
+           * We should explicitly disallow ET_DYN executables from having
+           * R_AARCH64_COPY relocations.
+           */
+         DL_ERR("%s R_AARCH64_COPY relocations only supported for ET_EXEC", name);
+         return -1;
+     }
+     count_relocation(kRelocCopy);
+     MARK(rela->r_offset);
+     TRACE_TYPE(RELO, "RELO COPY %16llx <- %lld @ %16llx %s\n",
+                reloc,
+                s->st_size,
+                (sym_addr + rela->r_addend),
+                sym_name);
+     if (reloc == (sym_addr + rela->r_addend)) {
+         ElfW(Sym)* src = soinfo_do_lookup(NULL, sym_name, &lsi);
+
+         if (src == NULL) {
+             DL_ERR("%s R_AARCH64_COPY relocation source cannot be resolved", name);
+             return -1;
+         }
+         if (lsi->has_DT_SYMBOLIC) {
+             DL_ERR("%s invalid R_AARCH64_COPY relocation against DT_SYMBOLIC shared "
+                    "library %s (built with -Bsymbolic?)", name, lsi->name);
+             return -1;
+         }
+         if (s->st_size < src->st_size) {
+             DL_ERR("%s R_AARCH64_COPY relocation size mismatch (%lld < %lld)",
+                    name, s->st_size, src->st_size);
+             return -1;
+         }
+         memcpy(reinterpret_cast<void*>(reloc),
+                reinterpret_cast<void*>(src->st_value + lsi->load_bias), src->st_size);
+     } else {
+         DL_ERR("%s R_AARCH64_COPY relocation target cannot be resolved", name);
+         return -1;
+     }
+     break;
       case R_AARCH64_TLS_TPREL64:
         TRACE_TYPE(RELO, "RELO TLS_TPREL64 *** %16llx <- %16llx - %16llx\n",
                    reloc, (sym_addr + rela->r_addend), rela->r_offset);
@@ -1465,17 +1501,48 @@ int soinfo::Relocate(ElfW(Rel)* rel, unsigned count) {
         *reinterpret_cast<ElfW(Addr)*>(reloc) += sym_addr - rel->r_offset;
         break;
       case R_ARM_COPY:
-        /*
-         * ET_EXEC is not supported so this should not happen.
-         *
-         * http://infocenter.arm.com/help/topic/com.arm.doc.ihi0044d/IHI0044D_aaelf.pdf
-         *
-         * Section 4.7.1.10 "Dynamic relocations"
-         * R_ARM_COPY may only appear in executable objects where e_type is
-         * set to ET_EXEC.
-         */
-        DL_ERR("%s R_ARM_COPY relocations are not supported", name);
-        return -1;
+     if ((flags & FLAG_EXE) == 0) {
+             /*
+              * http://infocenter.arm.com/help/topic/com.arm.doc.ihi0044d/IHI0044D_aaelf.pdf
+              *
+              * Section 4.7.1.10 "Dynamic relocations"
+              * R_ARM_COPY may only appear in executable objects where e_type is
+              * set to ET_EXEC.
+              *
+              * TODO: FLAG_EXE is set for both ET_DYN and ET_EXEC executables.
+              * We should explicitly disallow ET_DYN executables from having
+              * R_ARM_COPY relocations.
+              */
+             DL_ERR("%s R_ARM_COPY relocations only supported for ET_EXEC", name);
+             return -1;
+         }
+         count_relocation(kRelocCopy);
+         MARK(rel->r_offset);
+         TRACE_TYPE(RELO, "RELO %08x <- %d @ %08x %s", reloc, s->st_size, sym_addr, sym_name);
+         if (reloc == sym_addr) {
+             ElfW(Sym)* src = soinfo_do_lookup(NULL, sym_name, &lsi);
+
+             if (src == NULL) {
+                 DL_ERR("%s R_ARM_COPY relocation source cannot be resolved", name);
+                 return -1;
+             }
+             if (lsi->has_DT_SYMBOLIC) {
+                 DL_ERR("%s invalid R_ARM_COPY relocation against DT_SYMBOLIC shared "
+                        "library %s (built with -Bsymbolic?)", name, lsi->name);
+                 return -1;
+             }
+             if (s->st_size < src->st_size) {
+                 DL_ERR("%s R_ARM_COPY relocation size mismatch (%d < %d)",
+                        name, s->st_size, src->st_size);
+                 return -1;
+             }
+             memcpy(reinterpret_cast<void*>(reloc),
+                    reinterpret_cast<void*>(src->st_value + lsi->load_bias), src->st_size);
+         } else {
+             DL_ERR("%s R_ARM_COPY relocation target cannot be resolved", name);
+             return -1;
+         }
+         break;
 #elif defined(__i386__)
       case R_386_JMP_SLOT:
         count_relocation(kRelocAbsolute);
